### ISSUE TITLE
`exhaustive_cases` has misleading documentation and may need generalization

### ISSUE THREAD

**Issue by eernstg** [dart-lang/linter#2768]
The [documentation](https://dart-lang.github.io/linter/lints/exhaustive_cases.html) associated with the lint `exhaustive_cases` contains the following program:

```dart
class EnumLike {
  final int i;
  const E._(this.i);

  static const e = E._(1);
  static const f = E._(2);
  static const g = E._(3);
}

void bad(EnumLike e) {
  // Missing case.
  switch(e) { // LINT
    case E.e :
      print('e');
      break;
    case E.f :
      print('e');
      break;
  }
}
```

It should be adjusted to use `EnumLike` rather than `E` for every identifier spelled `E` in the program, to eliminate the errors. 

Next, a thing that doesn't matter, but doesn't look right either: It was probably the intention to do `print('f')` rather than `print('e')` at the second switch case (and `print('g')` at the third one in the 'GOOD' example).

Finally, the lint itself might benefit from a generalization: It should take all invocations of the `EnumLike` constructors into account (currently, it doesn't report the missing case issue if the declaration of `g` is moved to the top level).

It might be helpful to report that a given class is only _almost_ enum-like, because being non-enum-like will silently eliminate all the lint messages, and a developer who is relying on these messages may never notice that a given switch is non-exhaustive, but the expected lint message is never emitted just because `EnumLike` isn't enum-like enough.

Perhaps the enum-like class should be marked explicitly using `@enumLike` or something like that?

Or perhaps this lint should just be deprecated? It doesn't look like it's an easy job to make it work really well... ;-)

**Comment by bwilkerson** [comment]
Yes, the documentation needs to be fixed, thanks.

As for the other comments, I think that the key here is to do a better job of defining what we mean by "enum-like". In addition to the changes you noted, the documentation then needs to be updated to reflect the working definition.

Based on the user requests that led to this lint, one of the criteria (that's missing from the docs) we came up with is that the constructors are only invoked in the initializers of static fields of the class. Hence, invoking the constructor in the initializer of a top-level variable prevents the class from being considered enum-like.

By I do like the idea of having an annotation to mark classes that are intended to be used like an `enum` and to have additional checks to make sure they're used in a way that's consistent with that definition.

**Comment by Levi-Lesches** [comment]
> By I do like the idea of having an annotation to mark classes that are intended to be used like an enum and to have additional checks to make sure they're used in a way that's consistent with that definition.

I haven't heard much about the new `Enum` interface, other than seeing it in the [release notes](https://github.com/dart-lang/sdk/blob/master/CHANGELOG.md#dartcore). Would marking a class as `implements Enum` or something similar be the solution?

**Comment by bwilkerson** [comment]
Unfortunately, the class `Enum` can't be subtyped (except by an `enum` declaration).

**Comment by pq** [comment]
ps, @eernstg: thanks for catching all of these documentation gaffes.

I've got a fix out for those.  It'd be great to get some more input on the annotation idea (which appeals to me FWIW).

**Comment by pq** [comment]
Documentation should now be fixed. 👍
